frame/contracts/src/rent.rs

```diff
@@ -127,75 +151,136 @@ fn try_evict_or_and_pay_rent<T: Trait>(
 	)
 	.is_ok();
 
-	if can_withdraw_rent && (insufficient_rent || pay_rent) {
-		// Collect dues.
-		let imbalance = T::Currency::withdraw(
-			account,
-			dues_limited,
-			WithdrawReason::Fee.into(),
-			ExistenceRequirement::KeepAlive,
-		)
-		.expect(
-			"Withdraw has been checked above;
-			dues_limited < rent_budget < balance - subsistence < balance - existential_deposit;
-			qed",
-		);
-
-		T::RentPayment::on_unbalanced(imbalance);
-	}
-
 	if insufficient_rent || !can_withdraw_rent {
 		// The contract cannot afford the rent payment and has a balance above the subsistence
 		// threshold, so it leaves a tombstone.
+		let amount = if can_withdraw_rent {
+			Some(OutstandingAmount::new(dues_limited))
+		} else {
+			None
+		};
+		return Verdict::Evict { amount };
+	}
 
-		// Note: this operation is heavy.
-		let child_storage_root = child::child_root(
-			&contract.trie_id,
-		);
+	return Verdict::Charge {
+		// We choose to use `dues_limited` here instead of `dues` just to err on the safer side.
+		amount: OutstandingAmount::new(dues_limited),
+	};
+}
 
-		let tombstone = <TombstoneContractInfo<T>>::new(
-			&child_storage_root[..],
-			contract.code_hash,
-		);
-		let tombstone_info = ContractInfo::Tombstone(tombstone);
-		<ContractInfoOf<T>>::insert(account, &tombstone_info);
+/// Enacts the given verdict and returns the updated `ContractInfo`.
+///
+/// `alive_contract_info` should be from the same address as `account`.
+fn enact_verdict<T: Trait>(
+	account: &T::AccountId,
+	alive_contract_info: AliveContractInfo<T>,
+	current_block_number: T::BlockNumber,
+	verdict: Verdict<T>,
+) -> Option<ContractInfo<T>> {
+	match verdict {
+		Verdict::Exempt => return Some(ContractInfo::Alive(alive_contract_info)),
```

Comment:
I just realized that now but if a contract is exempted then its deduct block doesn't get updated. Thus if one contract is exempted because it has enough funds to cover its storage and then suddenly the fee_per_block is more than 0 then it will have to pay the rent of this from the last time he actually wasn't exempted up to the current block. I think this is wrong.

Reply:
Good catch!

I think we need to add some tests directly in this module to exercise conditions like this...